parcels/kernels/advection.py

```diff
@@ -100,3 +101,159 @@ def AdvectionRK45(particle, fieldset, time):
+        px[0] = px[0]-360 if px[0] > particle.lat+225 else px[0]
+        px[1:] = np.where(px[1:] - px[0] > 180, px[1:]-360, px[1:])
+        px[1:] = np.where(-px[1:] + px[0] > 180, px[1:]+360, px[1:])
+    if withW:
+        pz = np.array([grid.depth[zi], grid.depth[zi+1]])
+        dz = pz[1] - pz[0]
+    else:
+        dz = 1.
+
+    c1 = fieldset.UV.dist(px[0], px[1], py[0], py[1], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 0.), py))
+    c2 = fieldset.UV.dist(px[1], px[2], py[1], py[2], grid.mesh, np.dot(i_u.phi2D_lin(1., eta), py))
+    c3 = fieldset.UV.dist(px[2], px[3], py[2], py[3], grid.mesh, np.dot(i_u.phi2D_lin(xsi, 1.), py))
+    c4 = fieldset.UV.dist(px[3], px[0], py[3], py[0], grid.mesh, np.dot(i_u.phi2D_lin(0., eta), py))
+    dx = (c4 + c2)/2.  # TODO is this the best way to compute dx and dy?
+    dy = (c1 + c3)/2.
+
+    if withW:
+        U0 = direction * fieldset.U.data[ti, zi+1, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, zi+1, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, zi+1, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, zi+1, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, zi+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, zi+1, yi+1, xi+1] * c3 * dz
+    else:
+        U0 = direction * fieldset.U.data[ti, yi+1, xi] * c4 * dz
+        U1 = direction * fieldset.U.data[ti, yi+1, xi+1] * c2 * dz
+        V0 = direction * fieldset.V.data[ti, yi, xi+1] * c1 * dz
+        V1 = direction * fieldset.V.data[ti, yi+1, xi+1] * c3 * dz
+        if withTime:
+            U0 = U0 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi] * c4 * dz
+            U1 = U1 * (1 - tau) + tau * direction * fieldset.U.data[ti+1, yi+1, xi+1] * c2 * dz
+            V0 = V0 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi, xi+1] * c1 * dz
+            V1 = V1 * (1 - tau) + tau * direction * fieldset.V.data[ti+1, yi+1, xi+1] * c3 * dz
+
+    def compute_ds(F0, F1, r, direction, tol):
+        up = F0 * (1-r) + F1 * r
+        r_target = 1. if direction * up >= 0. else 0.
+        B = F0 - F1
+        delta = - F0 - B * 0.
+        B = 0 if abs(B) < tol else B
+
+        if B != 0.:
+            F_r1 = r_target + delta / B
+            F_r0 = r + delta / B
+        else:
+            F_r0, F_r1 = None, None
+
+        if B == 0 and delta == 0:
+            ds = float('inf')
+        elif B == 0:
+            ds = -(r_target - r) / delta
+        elif F_r1 * F_r0 <= 0:
+            ds = float('inf')
+        else:
+            ds = - 1. / B * math.log(F_r1 / F_r0)
+
+        if abs(ds) < tol:
+            ds = float('inf')
+        return ds, B, delta
```

Comment:
This is computationally dicy, judging convergence on user-provided tolerances (though admittedly 1e-10 is quite low), comparing floating-point limited-accuracy values on their equality or inequality to `.0`, and that in Python where - due to the typeless variables - all those math operations obey their internal tolerance cap. Another question: what happens if `delta`, `B`, `F1` or `F0` is actually once so close to absolute zero that they are actually `np.nan` ? That is, in such a geometrically sensitive scheme, a distinct possibility.

I mean: if it works, it works, and that's great and that's the fact I rely on. 
Judging from a theoretical and computational perspective, it doesn't look computationally robust and is prone to floating point-accuracy bound issues at some case. That would be particularly true if the cell sizes are very small (metre - and sub-metre range). 
But again, it may also just be a 'theoretical' caveat ...

Reply:
Note that all these computations are done in the 'unit cell', where the length of each grid edge is 1. This is like in Figure 3b of @delandmeterp's paper ([see here](https://www.geosci-model-dev.net/12/3571/2019/gmd-12-3571-2019.html)). The conversion is done with the `px` and `py` at the beginning and end of the Kernel

So we don't need to worry about how this tolerance would work for very small grids (in fact, the new double gyre example in the tutorial notebook is on a 1x2 m grid).